src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs

```diff
@@ -114,15 +114,24 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
 
             // Guarantee the existance of a context for every policy execution, but only create a new one if needed. This
             // allows later handlers to flow state if desired.
+            var cleanUpContext = false;
             var context = request.GetPolicyExecutionContext();
             if (context == null)
             {
                 context = new Context();
                 request.SetPolicyExecutionContext(context);
+                cleanUpContext = true;
             }
 
             var policy = _policy ?? SelectPolicy(request);
-            return policy.ExecuteAsync((c, ct) => SendCoreAsync(request, c, ct), context, cancellationToken);
+            var response = await policy.ExecuteAsync((c, ct) => SendCoreAsync(request, c, ct), context, cancellationToken);
+
+            if (cleanUpContext)
+            {
+                request.SetPolicyExecutionContext(null);
+            }
+
+            return response;
```

Comment:
Should this be in a try/finally?

Reply:
:+1: This isn't about releasing critical resources, but it does make the clean-up guarantee stronger.

Added this to the PR, with tests.  Also covered the case where _policySelector throws.